src/box/lua/config/applier/credentials.lua

```diff
@@ -277,6 +277,101 @@ local function privileges_add_defaults(name, role_or_user, intermediate)
     return res
 end
 
+-- The table contains lists of all actions per object, that should be performed
+-- after object creation. Actions are listed in the following format:
+-- [obj_type][obj_name] = {{
+--     role_or_user,
+--     grant_or_revoke,
+--     user_or_role_name,
+--     privileges
+--   }, {
+--     ...
+--   },
+-- }
+local actions_on_creation = {
+    ['space'] = {},
+    ['function'] = {},
+    ['sequence'] = {},
+}
```

Comment:
I see two possible ways to interpret a sequence of configs: snapshots or deltas. You can feel this distinction familiar by the [What is Git?](https://git-scm.com/book/en/v2/Getting-Started-What-is-Git%3F) section of Git Book.

While one representation is generally equivalent to another, there is a nuance. Let's consider an example of configuration snapshots: C1 -> C2 -> C3 and corresponding deltas: D1 (? -> C1), D2 (C1 -> C2), D3 (C2 -> C3).

Say, we have a function that grant/revoke needed privileges. It looks negligible, whether it accepts a snapshot (C3) or a delta (D3). The nuance here is that a delta is only valid, when it is applied on top of a right previous snapshot: applying D3 gives C3 if we're in C2, but may give some different C3' otherwise.

Back to our task. There are several points, where I'm in doubts about using deltas.

First, we're already operating by snapshots on startup/reload and on ro -> rw database mode change. The intermediate representation and all the functions around use it. Now, we have another delta-based intermediate representation and a parallel set of functions around it.

It seems, it reduces our ability to reuse the code and adds unnecessary entities.

I'm also in doubt about synchronization. If we would reuse the same code, we would reuse the synchronization code too.

Second, the delta is calculated using some current database schema state, while it is applied on a possibly different state. So, we can attempt to apply some changes twice (`if_not_exists` helps here) and, what is worse, we can miss some of the changes.

It seems that continue using snapshot-based approach should make the code simpler and more durable. If you considered this approach, please, share reasons to don't use it. Otherwise, please, pay some time to investigate this implementation variant.

Reply:
Implemented per-object sync inside the trigger.